shell/utils/resourceWatcher.js

```diff
@@ -0,0 +1,212 @@
+import Socket, {
```

Comment:
This needs some comments/docs about what this does and why

Reply:
added some comments, full docs should probably wait until the whole thing is implemented to avoid rewriting as much as possible.